Feature/search

AyaazStreet opened the pull request:
- Created new components for search page
- Created SearchPage
- Edited Navbar to change when on SearchPage
- Created fetch API call getSearch to fetch manga based of string query

AyaazStreet committed:
feat: :sparkles: Created SearchPage

Added a SearchPage and routes for it

AyaazStreet committed:
feat: :sparkles: Search

Created a search feature with formik and yup validation

theadley commented on src/api/categories.ts:
```diff
@@ -29,3 +29,10 @@ export async function getGenres(): Promise<Genre[]> {
+
+export async function getSearch(query: string): Promise<Comic[]> {
+    const res = await fetch(
+        `${baseUrl}v1.0/search/?type=comic&showall=true&q=${query}&t=true`
```
I want to make sure that `query` is URL encoded with `encodeURIComponent()` somewhere.

AyaazStreet replied:
I am right in saying just wrapping the query in `encodeURIComponent()` will ensure this. Like `${encodeURIComponent(query)}`

theadley replied:
Yes

theadley commented on src/components/Navbar.tsx:
```diff
@@ -1,9 +1,16 @@
+    search = false,
+    setSearchResult,
+}: {
+    search?: boolean
```
This is a weird name. Search doesn't sound like a boolean.

theadley commented on src/components/Navbar.tsx:
```diff
@@ -1,9 +1,16 @@
+    setSearchResult,
+}: {
+    search?: boolean
+    setSearchResult?: React.Dispatch<React.SetStateAction<string | undefined>>
```
This feels not good. What is this for and why cant it be a handler function where the parent sets the state instead of a SetStateAction

AyaazStreet replied:
I fixed this by removing it and using zustand to store search query.

theadley commented on src/components/Navbar.tsx:
```diff
@@ -56,26 +63,32 @@ function Navbar() {
-                    </svg>
-                </button>
+                {search ? (
+                    <SearchBar setSearchResult={setSearchResult!} />
```
Nice prop drilling going on here. Maybe time for some state management?

theadley replied:
Even if that's just context and useReduce

AyaazStreet replied:
I fixed this by removing it and using zustand to store search query.

theadley commented on src/pages/SearchPage.tsx:
```diff
@@ -0,0 +1,14 @@
+    const [searchResult, setSearchResult] = useState<string | undefined>()
+    return (
+        <>
+            <Navbar search={true} setSearchResult={setSearchResult}/>
```
Why have the prop if you hardcode it to true

AyaazStreet replied:
I did this to create a variation to my Navbar. When the optional search prop is true then the Navbar has a form element, otherwise it only has a button element.

theadley replied:
Is that used anywhere?
Also please rename it because as a person with no knowledge of that, how would I know what setting that boolean does.
Maybe `showSearchOnNav` or something more obvious

AyaazStreet replied:
It is used when you have the Navbar on `SearchPage.tsx`.
I have renamed it for more context.

theadley reviewed (changes requested):
Minor changes needed

AyaazStreet committed:
fix: :sparkles: PR Fixes

Used Zustand to store the state for searchQuery. Renamed Navbar prop. Removed prop drilling.

AyaazStreet committed:
refactor: :bulb: Removed Router Devtools

Commented out router Devtools for later use

theadley reviewed (approved):
Well done!
